### ISSUE TITLE
Fixed positioning

### ISSUE THREAD

**Issue by kusti8** [andlabs/libui#306]
Hi,

For `proton-native`, I want to incorporate Yoga Layout, which I can do on the JS side without adding extra dependencies here, but I would need a way to fixed position place widgets using x and y coordinates. I've been playing around with GTK, and I got it working here: https://github.com/kusti8/libui, but I don't know how to do that for Windows or for Mac. This would be more of a substitute for Grid and make it a lot easier to use and more stable.

Thanks,
Gustav

**Comment by andlabs** [comment]
Interesting that you did it with GTK+, as that's the one toolkit that already provides grids (and the uiGrid API is based on GtkGrid)... I'll look through all this (including yoga itself) later.

**Comment by cody271** [comment]
@kusti8,

Your `uiFixed` container control looks good to me, but `examples/layout-test/main.c` is missing in that commit. (Also your `uiFixedSignature` isn't valid ASCII, maybe use something like `0x46697864`, which is "Fixd".)

A while back, I actually looked into adding support for Yoga into libui as well. Though I gave up on it because the Yoga native C API is not documented at all. Only supporting Yoga from within JS would be fine I guess. Let me know if I can help on this.

**Comment by andlabs** [comment]
I've historically had problems with GtkFixed and background drawing under certain themes. In this case I'd just use a custom GtkContainer subclass to do all the work. I can write that myself, should I decide to have uiCustomContainer at all...

**Comment by kusti8** [comment]
Yeah I removed the test code but forgot to remove the options in the cake
file. I wasn't sure what the signature was, so I just made something up.
Didn't realize it was text. I didn't prep is as a PR, just as a quick proof
of concept.

Yeah so it seems like they used to have documentation, but removed it. If
you go back far enough in the commits they have some OK documentation. I
thought about adding it to libui but honestly I don't know enough about C
to link it and integrate it. If we add it in JS, then it doesn't clutter
libui, but it depends on whether you think it should be required/built in
for libui, or sort of optional.

I haven't looked at yoga's JS documentation, but I don't imagine it to be
too hard. My only problem is figuring out how to do a fixed view in Windows
and Mac. I think I got a rough outline done for Windows, but it's more
complex than GTK so I'm not totally sure.

**Comment by cody271** [comment]
@andlabs [`GtkFixed`](https://github.com/GNOME/gtk/blob/master/gtk/gtkfixed.c) is pretty straightforward. What would we actually do differently here?

@kusti8 Yoga would definitely be optional. My recommendation is to focus on getting Yoga from JS to work with your `uiFixed` and `libui-node` on just GTK+. Don't worry about the Mac and Windows implementations yet, we can flush out `uiFixed` once this all works and is useful.

**Comment by kusti8** [comment]
Yeah good idea. @parro-it any thoughts ? What do you think about integrating it into libui-node vs into proton-native?

**Comment by andlabs** [comment]
*is now considering calling libui's next build system "cake"*

@cody271 you will have to go back into the **really** early git history of package ui itself for the answers. I think I also have a Stack Overflow question on the issue as well.

Will respond to the rest later.

**Comment by parro-it** [comment]
> Yeah good idea. @parro-it any thoughts ? What do you think about integrating it into libui-node vs into proton-native?

I don't know yoga, I have to look at how it works... I saw there are some related packages on npm, maybe we can use them...

**Comment by parro-it** [comment]
Regarding where to put them, if there is a compilation phase required to use Yoga, I think libui-node could be a better choice, because otherwise we will have to prepare proton prebuilt binaries too.

**Comment by parro-it** [comment]
I saw yoga already have official bindings fir js, and they use nbind as libui-node does.

So, I think we can either use it as is on proton side, or include theirs cpp sources in libui-node and produce a single library.

**Comment by kusti8** [comment]
I've made a PR #332 with all of my changes, for Linux, Windows and Mac.
